quantum/deutsch_jozsa.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+"""
+Deutsch-Josza Algorithm is one of the first examples of a quantum
+algorithm that is exponentially faster than any possible deterministic
+classical algorithm
+
+Premise:
+We are given a hidden Boolean function f, 
+which takes as input a string of bits, and returns either 0 or 1:
+
+f({x0,x1,x2,...}) -> 0 or 1, where xn is 0 or 1
+ 
+The property of the given Boolean function is that it is guaranteed to
+either be balanced or constant. A constant function returns all 0's 
+or all 1's for any input, while a balanced function returns  0's for 
+exactly half of all inputs and 1's for the other half. Our task is to
+determine whether the given function is balanced or constant.
+
+References:
+- https://en.wikipedia.org/wiki/Deutsch-Jozsa_algorithm
+- https://qiskit.org/textbook/ch-algorithms/deutsch-jozsa.html
+"""
+
+import numpy as np
+import qiskit as q
+
+
+def dj_oracle(case: str, n: int) -> q.QuantumCircuit:
+    """
+    Returns a Quantum Circuit for the Oracle function.
+    The circuit returned can represent balanced or constant function,
+    according to the arguments passed
+    """
+    # This circuit has n+1 qubits: the size of the input,
+    # plus one output qubit
+    oracle_qc = q.QuantumCircuit(n + 1)
+
+    # First, let's deal with the case in which oracle is balanced
+    if case == "balanced":
+        # First generate a random number that tells us which CNOTs to
+        # wrap in X-gates:
+        b = np.random.randint(1, 2 ** n)
+        # Next, format 'b' as a binary string of length 'n', padded with zeros:
+        b_str = format(b, "0" + str(n) + "b")
+        # Next, we place the first X-gates. Each digit in our binary string
+        # correspopnds to a qubit, if the digit is 0, we do nothing, if it's 1
+        # we apply an X-gate to that qubit:
+        for qubit in range(len(b_str)):
```

Comment:
Please use enumerate() instead of range(len()).

Reply:
qubit is not a qubit, it is an int.  Maybe call it I or index.

Reply:
using _enumerate_ now, changed var name to _index_